### PR TITLE
core/credential,mount: re-read /proc/self/mountinfo before invoking umount command

### LINKED ISSUE #25527: Umount failure in mkosi and integration test CI
### systemd version the issue has been seen with

main

### Used distribution

_No response_

### Linux kernel version used

_No response_

### CPU architectures issue was seen on

_No response_

### Component

systemd-tmpfiles

### Expected behaviour you didn't see

_No response_

### Unexpected behaviour you saw

_No response_

### Steps to reproduce the problem

_No response_

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
[  365.166858] umount[63]: umount: /run/credentials/systemd-tmpfiles-setup.service: no mount point specified.
run-credentials-systemd\x2dfirstboot.service.mount: Deactivated successfully.
[  OK  ] Unmounted /run/credentials/systemd-firstboot.service.
run-credentials-systemd\x2dsysusers.service.mount: Deactivated successfully.
[  OK  ] Unmounted /run/credentials/systemd-sysusers.service.
run-credentials-systemd\x2dtmpfiles\x2dsetup.service.mount: Mount process exited, code=exited, status=32/n/a
run-credentials-systemd\x2dtmpfiles\x2dsetup.service.mount: Failed with result 'exit-code'.
```
```

### PR DESCRIPTION
**Pull request by yuwata**
When a unit has credentials, stopping the service unmounts the credentials
directory. On shutdown, stopping the service and the corresponding mount
unit may be done mostly simultaneously, and if we invoke umount command soon
after umount() being called on stopping the service, the mount unit will
fail.

This makes Mount.dirty flag set when umount() is called for a path, and
re-read /proc/self/mouninfo before invoking umount command if the flag is set.

Fixes #25527.

This also reverts several commits in #28787, and addresses the post-merge comments for the PR.

### CODE REVIEW

**Comment by keszybz** [comment]
Hmm, is it useful at all to have mount units for the credentials directories? We don't create mount units for API directories (`/dev/`, `/sys`, `/proc`, etc.). The lifetime of the credentials mountpoint is tied to the service. As credentials become more widely used, we'll end up with thousands of mount units that are not useful for anything. I would prefer if we just get rid of them.

**Comment by yuwata** [comment]
> Hmm, is it useful at all to have mount units for the credentials directories? We don't create mount units for API directories (`/dev/`, `/sys`, `/proc`, etc.). The lifetime of the credentials mountpoint is tied to the service. As credentials become more widely used, we'll end up with thousands of mount units that are not useful for anything. I would prefer if we just get rid of them.

That make some sense for me, but out of scope of this PR. I'd like to address that if @poettering agree with that.

File: src/core/mount.c [hunk 1]
```diff
@@ -1284,6 +1284,11 @@ static int mount_stop(Unit *u) {
 
         assert(m);
 
+        /* When we directly call umount() for a path, then the state of the corresponding mount unit may be
+         * outdated. Let's re-read mountinfo now and update the state. */
+        if (m->dirty)
```
Comment (poettering) [inline src/core/mount.c]: i think "dirty" is the wrong term here, that generally indicates data that is not flushed to disk yet. But here we want to mark a state as invalidated.

Hence maybe call this "invalidated"? or "invalidated_state"?

**Comment by poettering** [comment]
I think this could work, one bikeshedding comment on the naming though

**Comment by poettering** [comment]
> Hmm, is it useful at all to have mount units for the credentials directories? We don't create mount units for API directories (`/dev/`, `/sys`, `/proc`, etc.). The lifetime of the credentials mountpoint is tied to the service. As credentials become more widely used, we'll end up with thousands of mount units that are not useful for anything. I would prefer if we just get rid of them.

dunno, i don't see "thousands" showing up just yet. I think people are already surprised that we exclude some mounts from getting .mount units, and I think we should probably not generate confusion around this needlessly.

i'd probably not change this behaviour unless this really becomes an issue.

I mean, to turn this around the ability to schedule unmount of these mounts (because they are mount units) might actually be useful, since we it might make sense to order them against swap detaching, dunno.

File: src/core/mount.c (hunk 1 above)
Comment (yuwata) [reply src/core/mount.c]: Renamed to `invalidated_state`.

**Pull request merged**